### ISSUE TITLE
abstract WordPress_Sniff

### ISSUE THREAD

**Issue by GaryJones** [WordPress-Coding-Standards/WordPress-Coding-Standards#246]
As per my 5th point of https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/pull/243#issuecomment-56791336, if we're going to have more sniffs bypassable via end-of-line whitelisting comments, we should create a way for all our sniffs to access that feature, without duplicating the method to do it.

We can do that in an abstract class that sits implements `PHP_CodeSniffer_Sniff` then update our sniffs which use the features to extend our new abstract class.

I propose something like:

~~~php
abstract class WordPress_Sniff implements PHP_CodeSniffer_Sniff {
	/**
	 * Get the last pointer in a line
	 *
	 * @since 1.0.0
	 *
	 * @param array   $tokens   Tokens.
	 * @param integer $stackPtr The position of the current token in the	stack passed in $tokens.
	 *
	 * @return integer Next pointer position.
	 */
	protected function get_last_ptr( $tokens, $stackPtr ) {
		$currentLine = $tokens[$stackPtr]['line'];
		$nextPtr = $stackPtr;
		while ( isset( $tokens[$nextPtr + 1]['line'] ) && $tokens[$nextPtr + 1]['line'] == $currentLine ) {
			$nextPtr++;
			// Do nothing, we just want the last token of the line.
		}

		return $nextPtr;
	}
	
	/**
	 * Find whitelisting comment.
	 *
	 * Comment must be at the end of the line, and use // format.
	 * It can be prefixed or suffixed with anything e.g.
	 * ... // foobar okay
	 * ... // WPCS: foobar okay.
	 * are both fine, the latter meeting starting capital and trailing full stop
	 * that single comments may require.
	 *
	 * @since ...
	 *
	 * @param string  $comment  Comment to find.
	 * @param array   $tokens   Tokens.
	 * @param integer $stackPtr The position of the current token in the stack passed in $tokens.
	 *
	 * @return boolean True if whitelisting comment was found, false otherwise.
	 */
	protected function has_whitelist_comment( $comment, array $tokens, $stackPtr ) {
		$lastPtr = $this->get_last_ptr( $tokens, $stackPtr );

		return $tokens[$lastPtr]['code'] === T_COMMENT &&
			preg_match( '#' . preg_quote( $comment ) .'#i', $tokens[$lastPtr]['content'] ) > 0
		);
	}
}

// Change sniffs that need the functionality to
// extend WordPress_Sniff instead of PHP_CodeSniffer_Sniff
class WordPress_Sniffs_PHP_FoobarSniff extends WordPress_Sniff {
	function process( PHP_CodeSniffer_File $phpcsFile, $stackPtr )
	{
		$tokens = $phpcsFile->getTokens();

		if ( ! $this->has_whitelist_comment( 'foobar okay', $tokens, $stackPtr ) {
			$error = 'Found: ' . $tokens[$stackPtr]['content'] . '. Use foobar.';
			$phpcsFile->addWarning($error, $stackPtr);
		}
	}
}
~~~

Untested, but you should get the idea.

As you can see, the `WordPress_Sniff` class also includes any other common patterns that occur over our sniffs - not just finding if a comment is the last thing on a line, but a method for getting the last pointer, etc. Of itself, these methods should be fairly easy to be unit testable too.

**Comment by JDGrimes** [comment]
I was working on a PR for this, and here's what I'm thinking. In the above proposal, a bootstrap for performing various searches (and other operations) on the tokens is provided by custom methods of the abstract sniff class. While this reduces duplication across our sniffs, it also requires some amount of duplication across these methods. In each of the above methods, a `$tokens` parameter is required. Picture a bunch more methods. Every one will require that parameter. So we'll be passing the `$tokens` from method to method in a class. That's silly. We're in a class, so a value that needs to be accessed by multiple methods should be a class property. Unfortunately, the design of PHPCS is such that we can't automatically set the tokens as a class property. It can't be bootstrapped, but would have to be done manually by every sniff that uses these methods.

PHPCS's design actually makes this kind of bootstrapping difficult. What token-handling bootstrap is provided, is provided by the `PHP_CodeSniffer_File` class, which is passed to the sniffs' process methods as `$phpcsFile`. Any custom bootstrap like this will probably need the `$phpcsFile` being processed so that it can utilize `findNext()`, etc. (This proved true when adapting @GaryJones proposal to real life; instead of `$tokens`, we'll need to pass `$phpcsFile` around.)

So, any thoughts on requiring each child sniff to do something like this:

```php
$this->phpcsFile = $phpcsFile;
```

Or this:

```php
$this->initFile( $phpcsFile );
```

Or maybe someone has a better idea?
